_posts/2018-10-08-http-part5.markdown

```diff
@@ -0,0 +1,305 @@
+[It seems](https://stackoverflow.com/a/10241044/1189159) that there is no information about `FIN,ACK` event from the client in Socket API.
+The only indication of it is to try to either use `read()` or `readLine()` on `BufferedReader`.
+
+With that in mind, it is hard to do the following logic - block until new data arrives, then parse the request, if there is no data, close a connection. We cannot use `readLine()` before `Request.fromRaw(input)` because it will consume one line of our `BufferedReader`. We could change signature `fromRaw` to accept the first line as `String` and the rest of a stream as `BufferedReader` but that’s just ugly.
+That’s why for now we can keep things as they were. The beginning of our parsing logic protects us from empty requests anyways.
+``` kotlin
+val requestLine = input.readLine() ?: throw EmptyRequestException()
+```
+If `readLine()` returns null, we can just catch it (7) and close the connection. It might not be the best, it may smell like using exceptions for control flow, but it works. 
+We will try to tackle this problem in the future maybe by using some sort of “one line/character caching BufferedReader”.
+
+## Tests
+The main happy path to test is to check whether we can make several requests on one connection.
+``` kotlin
+private val client = HttpUrlConnectionClient(connectionTimeout = 50, socketTimeout = 100)
+
+@Test
+fun `should reuse connection to make requests`() {
+    // given
+    server = Server(handler = {
+        Response(status = Status.OK, headers = HttpHeaders("content-length" to "0"))
+    })
+    server.start(8090)
+
+    // when
+    repeat(times = 3) {
+        client.exchange(url = "http://localhost:8090", request = Request(path = "/", method = RequestMethod.GET))
+    }
+
+    // then
+    assertThat(server.connectionsEstablished()).isEqualTo(1)
+}
+```
+To test that, we have to expose `connectionsEstablished()` method that returns the count of connections established to the server. This most likely will be transformed in some kind of `ServerStats` object, because we will want to have metrics, but let’s keep it simple for now.
+
+### Importance of content-length header
+When we switched to the persistent connection model, it is really important to send a `content-length` header even if we don't send body! Previously we would send a response and close the connection. Even if a client read body and didn’t know that it ended, it didn’t block waiting for more for a long time, because we immediately closed connection on the server side. Now have to explicitly say how long the body is because we don’t close the connection immediately.
```

Comment:
it didn’t block waiting for more content for a long time

Reply:
Now (who?) have to explicitly say how long the body is because we don’t close the connection immediately.

I guess in brackets should be 'we', as 'we have to say sth...'